pkg/worker/worker.go

```diff
@@ -17,79 +14,139 @@ import (
+	w.consumer = consumer
 
-var ErrNoDeviceMatch = errors.New("no device in device service matches EUI of uplink")
+	return &w
+}
 
-func startConsuming(c <-chan amqp091.Delivery, process Processor, p chan<- mq.PublishMessage) {
-	consume := func(delivery amqp091.Delivery) error {
-		var err error
-		var msg pipeline.Message
-		if err := json.Unmarshal(delivery.Body, &msg); err != nil {
-			return fmt.Errorf("could not unmarshal delivery: %v", err)
+// Will run the given processor. Any returned message will be sent to it's next defined step in the pipeline
+func (w *worker) Run() {
+	// Await any messages that appear on the message queue
+	for delivery := range w.consumer {
+		var incoming pipeline.Message
+		if err := json.Unmarshal(delivery.Body, &incoming); err != nil {
+			log.Printf("Error converting delivery: %v\n", err)
+			w.publishError(err)
+			delivery.Nack(false, false)
+			continue
 		}
 
-		// Do process
-		msg, err = process(msg)
+		// Once a message has been received, process it using the worker-unique processor
+		result, err := w.processor(incoming)
 		if err != nil {
-			return fmt.Errorf("could not process message: %v", err)
+			log.Printf("Error processing delivery: %v\n", err)
+			w.publishError(err)
+			delivery.Nack(false, false)
+			continue
 		}
 
-		// Publish result
-		topic, err := msg.NextStep()
+		// If the worker succesfully processed the result, publish it to the next message queue
+		topic, err := incoming.NextStep()
 		if err != nil {
-			return fmt.Errorf("message has no steps remaining: %w", err)
+			delivery.Nack(false, false)
+			w.publishError(err)
+			continue
 		}
-		msgJSON, err := json.Marshal(msg)
+		msgJSON, err := json.Marshal(result)
 		if err != nil {
-			return fmt.Errorf("could not marshal pipelines message: %w", err)
+			delivery.Nack(false, false)
+			w.publishError(fmt.Errorf("could not marshal pipelines message: %w", err))
+			continue
 		}
-		p <- mq.PublishMessage{Topic: topic, Publishing: amqp091.Publishing{
+		w.publisher <- mq.PublishMessage{Topic: topic, Publishing: amqp091.Publishing{
 			Body: msgJSON,
 		}}
-		return nil
-	}
 
-	for delivery := range c {
-		if err := consume(delivery); err != nil {
-			log.Printf("Error processing delivery: %v\n", err)
-			delivery.Nack(false, false)
-			continue
-		}
+		// The message was succesfully handled, ack the message.
 		delivery.Ack(false)
 	}
+
+	// Shutdown the MQ connection
+	w.cancelToken <- true
 }
+
+func (w *worker) publishError(err error) error {
+	errJSON, err := json.Marshal(workerError{
+		Origin:     w.appName,
+		OriginType: w.appType,
+		Error:      err.Error(),
+	})
+	if err != nil {
+		return fmt.Errorf("could not marshal json: %w", err)
+	}
+	w.publisher <- mq.PublishMessage{Topic: w.mqErrTopic, Publishing: amqp091.Publishing{
+		Body: errJSON,
+	}}
+	return nil
+}
+
+type workerError struct {
+	Origin     string `json:"origin"`
+	OriginType string `json:"originType"`
+	Error      string `json:"error"`
```

Comment:
AppName / Type not used in prod. Perhaps include the Message ID and the topic which was used for incoming. Though an identifier for the worker would be nice to have as well

Reply:
Perhaps use:

Worker name + Worker version + hostname

hostname is an env variable `env.must("HOSTNAME")`